links/link_vxlan.go

```diff
@@ -0,0 +1,358 @@
+)
+
+// LinkVxlanRaw is the raw (string) representation of a vxlan link as defined in the topology file.
+type LinkVxlanRaw struct {
+	LinkCommonParams `yaml:",inline"`
+	Remote           string      `yaml:"remote"`
+	VNI              int         `yaml:"vni"`
+	Endpoint         EndpointRaw `yaml:"endpoint"`
+	UDPPort          int         `yaml:"udp-port,omitempty"`
+	ParentInterface  string      `yaml:"parent-interface,omitempty"`
+	NoLearning       bool        `yaml:"no-learning,omitempty"`
+	NoL2Miss         bool        `yaml:"no-l2miss,omitempty"`
+	NoL3Miss         bool        `yaml:"no-l3miss,omitempty"`
+
+	// we use the same struct for vxlan and vxlan stitch, so we need to differentiate them in the raw format
+	LinkType LinkType
+}
+
+func (lr *LinkVxlanRaw) Resolve(params *ResolveParams) (Link, error) {
+	switch lr.LinkType {
+	case LinkTypeVxlan:
+		return lr.resolveVxlan(params)
+
+	case LinkTypeVxlanStitch:
+		return lr.resolveStitchedVxlan(params)
+
+	default:
+		return nil, fmt.Errorf("unexpected LinkType %s for Vxlan based link", lr.LinkType)
+	}
+}
+
+func (lr *LinkVxlanRaw) resolveStitchedVxlanComponent(params *ResolveParams) (*LinkVxlan, error) {
+	var err error
+
+	link := &LinkVxlan{
+		LinkCommonParams: lr.LinkCommonParams,
+		noLearning:       lr.NoLearning,
+		noL2Miss:         lr.NoL2Miss,
+		noL3Miss:         lr.NoL3Miss,
+	}
+
+	// point the vxlan endpoint to the host system
+	vxlanRawEp := lr.Endpoint
+	vxlanRawEp.Iface = fmt.Sprintf("vx-%s_%s", lr.Endpoint.Node, lr.Endpoint.Iface)
+
+	if params.VxlanIfaceNameOverwrite != "" {
+		vxlanRawEp.Iface = fmt.Sprintf("vx-%s", params.VxlanIfaceNameOverwrite)
+	}
+
+	// in the stiched vxlan mode we create vxlan interface in the host node namespace
+	vxlanRawEp.Node = "host"
+	vxlanRawEp.MAC = ""
+
+	// resolve local Endpoint
+	link.localEndpoint, err = vxlanRawEp.Resolve(params, link)
+	if err != nil {
+		return nil, err
+	}
+
+	ip := net.ParseIP(lr.Remote)
+	// if the returned ip is nil, an error occured.
+	// we consider, that we maybe have a textual hostname
+	// e.g. dns name so we try to resolve the string next
+	if ip == nil {
+		ips, err := net.LookupIP(lr.Remote)
+		if err != nil {
+			return nil, err
+		}
+
+		// prepare log message
+		sb := strings.Builder{}
+		for _, ip := range ips {
+			sb.WriteString(", ")
+			sb.WriteString(ip.String())
+		}
+		log.Debugf("looked up hostname %s, received IP addresses [%s]", lr.Remote, sb.String()[2:])
+
+		// always use the first address
+		if len(ips) <= 0 {
+			return nil, fmt.Errorf("unable to resolve %s", lr.Remote)
+		}
+		ip = ips[0]
+	}
+
+	parentIf := lr.ParentInterface
+
+	if parentIf == "" {
+		conn, err := rtnl.Dial(nil)
+		if err != nil {
+			return nil, fmt.Errorf("can't establish netlink connection: %s", err)
+		}
+		defer conn.Close()
+		r, err := conn.RouteGet(ip)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find a route to VxLAN remote address %s", ip.String())
+		}
+		parentIf = r.Interface.Name
+	}
```

Comment:
@steiler I think this code piece is now shared between 
`func (lr *LinkVxlanRaw) resolveStitchedVxlanComponent(params *ResolveParams) (*LinkVxlan, error)`

and 

`func (lr *LinkVxlanRaw) resolveVxlan(params *ResolveParams) (Link, error)`

It is a good candidate to create two common funcs in utils.

Reply:
No not in utils but still in LinkVxlanRaw.
However there is a slight difference in the vxlan endpoint part.
Anyways, let me see how we can carve the common part out.

Reply:
@hellt, take a look at 8458f0b